sympy/polys/matrices/sdm.py

```diff
@@ -0,0 +1,432 @@
+
+from .ddm import DDM
+
+
+class SDM(dict):
+    """Sparse matrix based on polys domain elements
+
+    This is a dict subclass and is a wrapper for a dict of dicts that supports
+    basic matrix arithmetic +, -, *, **.
+    """
+    def __init__(self, elemsdict, shape, domain):
+        super().__init__(elemsdict)
+        self.shape = self.rows, self.cols = m, n = shape
+        self.domain = domain
+
+        if not all(0 <= r < m for r in self):
+            raise DDMBadInputError("Row out of range")
+        if not all(0 <= c < n for row in self.values() for c in row):
+            raise DDMBadInputError("Column out of range")
+
+    @classmethod
+    def new(cls, sdm, shape, domain):
+        return cls(sdm, shape, domain)
+
+    def copy(A):
+        Ac = {i: Ai.copy() for i, Ai in A.items()}
+        return A.new(Ac, A.shape, A.domain)
+
+    @classmethod
+    def from_list(cls, ddm, shape, domain):
+        m, n = shape
+        if not (len(ddm) == m and all(len(row) == n for row in ddm)):
+            raise DDMBadInputError("Inconsistent row-list/shape")
+        getrow = lambda i: {j:ddm[i][j] for j in range(n) if ddm[i][j]}
+        irows = ((i, getrow(i)) for i in range(m))
+        sdm = {i: row for i, row in irows if row}
+        return cls(sdm, shape, domain)
+
+    @classmethod
+    def from_ddm(cls, ddm):
+        return cls.from_list(ddm, ddm.shape, ddm.domain)
+
+    def to_list(M):
+        m, n = M.shape
+        zero = M.domain.zero
+        ddm = [[zero] * n for _ in range(m)]
+        for i, row in M.items():
+            for j, e in row.items():
+                ddm[i][j] = e
+        return ddm
+
+    def to_ddm(M):
+        return DDM(M.to_list(), M.shape, M.domain)
+
+    @classmethod
+    def zeros(cls, shape, domain):
+        return cls({}, shape, domain)
+
+    @classmethod
+    def eye(cls, size, domain):
+        one = domain.one
+        sdm = {i: {i: one} for i in range(size)}
+        return cls(sdm, (size, size), domain)
+
+    def transpose(M):
+        MT = {}
+        for i, Mi in M.items():
+            for j, Mij in Mi.items():
+                try:
+                    MT[j][i] = Mij
+                except KeyError:
+                    MT[j] = {i: Mij}
+        return M.new(MT, M.shape[::-1], M.domain)
+
+    def __mul__(a, b):
+        if b in a.domain:
+            return a.mul(b)
+        else:
+            return NotImplemented
+
+    def __rmul__(a, b):
+        if b in a.domain:
+            return a.mul(b)
+        else:
+            return NotImplemented
+
+    def matmul(A, B):
+        if A.domain != B.domain:
+            raise DDMDomainError
+        C = {}
+        BT = B.transpose()
+        for i, Ai in A.items():
+            Ai_ks = set(Ai)
+            Ci = {}
+            for j, BTj in BT.items():
+                BTj_ks = set(BTj)
+                knonzero = Ai_ks & BTj_ks
+                if knonzero:
+                    Cij = sum(Ai[k] * BTj[k] for k in knonzero)
```

Comment:
Maybe `A.domain.zero` could be added as the start parameter.

Reply:
Yes, good point